libraries/AP_HAL/Util.cpp

```diff
@@ -77,6 +77,23 @@ uint64_t AP_HAL::Util::get_system_clock_ms() const
 #endif
 }
 
+uint64_t AP_HAL::Util::get_system_clock_us() const
+{
+#if defined(__APPLE__) && defined(__MACH__)
+    struct timeval ts;
+    gettimeofday(&ts, nullptr);
+    return ((long long)((ts.tv_sec * 1000000) + ts.tv_usec));
+#elif CONFIG_HAL_BOARD == HAL_BOARD_CHIBIOS
+    return ST2US(chVTGetSystemTime());
```

Comment:
this is bad in several ways on ChibiOS
First off, its calling a ChibiOS API that is only allowed in normal thread context from a function that users may think is OK from within a lock context or ISR. That could cause a fault (and hard crash)
Secondly, system ticks wrap really fast. So the result will wrap every few seconds on some boards.
Again, I think this should be in a AP_RTClock library, with the HAL providing the minimum information needed. Maybe a hal.util->get_initial_system_time(uint64_t &system_time_us, uint64_t &time_since_boot_us);
that would give two 64 bit timestamps, one in time since boot, the other a real time (since the epoch). It should return a bool, and return false if the system can't provide that time.

Reply:
Again, this code is exactly the same that exist in `uint64_t AP_HAL::Util::get_system_clock_ms()`, only converting to us. It's working with pixhawk and I believe that will work with other boards. I believe that's not necessary to hold this PR If modifications in chibios or the creation of a new library is necessary.